ykrt/src/mt.rs

```diff
@@ -347,12 +355,76 @@ impl MTThread {
                     }
                 }
             }
-            PHASE_TRACING => {
+            PHASE_TRACING | PHASE_TRACING_LOCK => {
+                if lp & PHASE_TAG == PHASE_TRACING_LOCK {
+                    // PHASE_TRACING_LOCK requires special handling, because another thread might
+                    // have grabbed it on this Location even though this thread is tracing this
+                    // location.
+
+                    if self.inner.tracer.is_none() {
+                        // If this thread isn't tracing anything, there's little point in spinning
+                        // as the thread tracing this location is still plausibly alive.
+                        return None;
+                    }
+
+                    // We wait to grab the lock.
+                    // FIXME: this is a terrible spinlock.
+                    while lp & PHASE_TAG == PHASE_TRACING_LOCK {
+                        lp = loc.state.load(Ordering::Acquire);
```

Comment:
Should we swap this line with the next? By yielding between reading and checking the value, we are maximising the chance of the information being stale, are we not?

Reply:
True. This doesn't affect correctness, but it's clearly inefficient. Fixed in https://github.com/softdevteam/yk/pull/199/commits/c6f0a220f9d9565164854b81bcf5347c161ebe93.